network_simulator/plot_tool.py

```diff
@@ -0,0 +1,390 @@
+        iterature through the data field of the tuples and outputs the
+        average value to std out.
+
+        :param string id: the Device of this average.
+        :param string value_type: the data type of the average.
+        :param string units: units of the average. e.g. "(sec)"
+        :param list tuple_list: the tuple list to be computed.
+        :return: no return, just print out the statement to stdout.
+        """
+        count = len(tuple_list)
+        avg = 0
+        if count != 0:
+            latest_timestamp = tuple_list[len(tuple_list) - 1][0]
+            avg = float(count) / latest_timestamp
+        print id + " Average " + value_type + ": " + str(avg) + " " + units
+
+
+
+class Grapher(object):
+    """
+    Graphs statistics items from Statistics class.
+    """
+    def __init__(self, stats=None):
+        """
+        :ivar Statistics stats: input Statistics object for graphing.
+        """
+        self.stats = stats
+
+    def graph_links(self):
+        """
+        Graphs the link stats of each available link.
+
+        This includes buffer occupancy, packet losses, and packet transmitted
+        (flow rate).
+        """
+        # Set up legend storage for link names.
+        # Buffer occupancy legend.
+        occpy_legend = []
+        # Packet loss legend.
+        loss_legend = []
+        # Packet transmission legend.
+        trans_legend = []
+
+        # Iterate through each link to display each stats.
+        for link_name, link_stats in self.stats.link_stats.iteritems():
+            # Name of link would be "link_name".
+            # Buffer occupancy w.r.t time.
+            plt.subplot(431)
+            PlotTool().graph_tuple_list(link_stats.buffer_occupancy)
+            occpy_legend.append(link_name)
+            # Output average buffer occupancy to stdout.
+            PlotTool().output_sum_avg_tuple_list(
+                "Link " + link_name,
+                "Buffer Occupancy",
+                True, "(pkts)", link_stats.buffer_occupancy)
+
+            # Packet loss times.
+            plt.subplot(434)
+            loss_count_lst = PlotTool().gen_count_interval_list(
+                link_stats.packet_loss_times)
+            PlotTool().graph_tuple_list(loss_count_lst)
+            loss_legend.append(link_name)
+            # Output average packet loss per sec to stdout.
+            PlotTool().output_count_avg_tuple_list(
+                "Link " + link_name,
+                "Packet Loss", "(pkts / sec)",
+                link_stats.packet_loss_times)
+
+            # Packet transmission w.r.t time.
+            plt.subplot(437)
+            flow_rate_list = PlotTool().gen_rate_tuple_list(
+                link_stats.packet_transmit_times)
+            PlotTool().graph_tuple_list(flow_rate_list)
+            trans_legend.append(link_name)
+            # Output average packet transmitted per sec to stdout.
+            PlotTool().output_count_avg_tuple_list(
```

Comment:
We actually want to output the average flow rate on the link in Mbps. To do this, you need to just sum up all of the data sent (in megabits) and divide by the biggest timestamp in order to get a rate in bps. Then divide by MEGABIT.